File: shaperglot/checks/no_orphaned_marks.py
from functools import lru_cache

from youseedee import ucd_data

from shaperglot.checks.orthographies import OrthographiesCheck

from .common import shaping_input_schema, ShaperglotCheck, check_schema


@lru_cache(maxsize=None)
def _simple_mark_check(codepoint):
    return ucd_data(codepoint).get("General_Category") == "Mn"


class NoOrphanedMarksCheck(ShaperglotCheck):
    name = "no_orphaned_marks"
    schema = check_schema(
        {
            "input": shaping_input_schema,
        }
    )

    def describe(self):
        return f"that, when {self.input.describe()}, no marks are left unattached"

    def execute(self, checker):
        buffer = self.input.shape(checker)
        dotted_circle_glyph = checker.cmap.get(0x25CC)

        # GDEF may be wrong, don't trust it.
        passed = True
        previous = None
        # pylint: disable=C0103
        for ix, (info, pos) in enumerate(
            zip(buffer.glyph_infos, buffer.glyph_positions)
        ):
            glyphname = checker.glyphorder[buffer.glyph_infos[ix].codepoint]
            # Is this a mark glyph?
            if info.codepoint == 0:
                passed = False
                message = "Shaper produced a .notdef"
                # Generally this message happens because we're missing
                # base/mark glyphs, so we will heard about this already,
                # *except* if no exemplars are defined, when it would be
                # good to hear specifically what glyphs we tried to test here.
                if not checker.lang.get("exemplarChars"):
                    message += " when shaping " + self.input.describe()

                checker.results.fail(
                    check_name="no-orphaned-marks",
                    result_code="notdef-produced",
                    message=message,
                    context={"text": self.input.check_yaml},
                )
                break
            if _simple_mark_check(checker.codepoint_for(glyphname)):
                # Was the previous glyph dotted circle?
                if previous and previous == dotted_circle_glyph:
                    passed = False
                    checker.results.fail(
                        check_name="no-orphaned-marks",
                        result_code="dotted-circle-produced",
                        message="Shaper produced a dotted circle when shaping "
                        + self.input.describe(),
                        context={"text": self.input.check_yaml},
                    )
                elif pos.x_offset == 0 and pos.y_offset == 0:  # Suspicious
                    passed = False
                    checker.results.fail(
                        check_name="no-orphaned-marks",
                        result_code="orphaned-mark",
                        message=f"Shaper didn't attach {glyphname} to {previous}",
                        context={
                            "text": self.input.check_yaml,
                            "mark": glyphname,
                            "base": previous,
                        },
                    )
            previous = glyphname
        if passed:
            checker.results.okay(
                check_name="no-orphaned-marks",
                message="No unattached mark glyphs were produced "
                + self.input.describe(),
                context={"text": self.input.check_yaml},
            )


class NoOrphanedMarksInOrthographiesCheck(NoOrphanedMarksCheck):
    name = "no_orphaned_marks_in_orthographies"

    def __init__(self, lang):
        super().__init__({
            "input": {
                "text":" ".join(OrthographiesCheck(lang).bases)
            }
        })

File: shaperglot-lib/src/checks/no_orphaned_marks.rs
use super::CheckImplementation;
use crate::{
    checker::Checker,
    reporter::{Fix, Problem},
    shaping::ShapingInput,
};
use itertools::Itertools;
use serde::{Deserialize, Serialize};
use unicode_properties::{GeneralCategory, UnicodeGeneralCategory};

#[derive(Serialize, Deserialize, Debug, Clone)]
/// A check implementation which ensures marks are anchors to their respective base characters
pub struct NoOrphanedMarks {
    /// The strings to shape and check
    test_strings: Vec<ShapingInput>,
    /// Whether the language has orthography data
    ///
    /// If this is true, we will not report notdefs, as the orthography check will
    /// catch them.
    has_orthography: bool,
}

impl CheckImplementation for NoOrphanedMarks {
    fn name(&self) -> String {
        "No Orphaned Marks".to_string()
    }

    fn should_skip(&self, _checker: &Checker) -> Option<String> {
        None
    }

    fn execute(&self, checker: &Checker) -> (Vec<Problem>, usize) {
        let tests_run = self.test_strings.len();
        let dotted_circle = checker.cmap.get(&0x25CC).cloned();
        let mut problems = vec![];

        for string in self.test_strings.iter() {
            let mut previous = None;
            let literally_a_dotted_circle = string.text.chars().any(|c| c == '\u{25CC}');
            let glyph_buffer = string
                .shape(checker)
                .expect("Failed to shape string for NoOrphanedMarks");
            for (codepoint, position) in glyph_buffer
                .glyph_infos()
                .iter()
                .zip(glyph_buffer.glyph_positions().iter())
            {
                // We got a notdef. The orthographies check would tell us about missing
                // glyphs, so if we are running one (we have exemplars) we ignore it; if not,
                // we report it.
                if codepoint.glyph_id == 0 && !self.has_orthography {
                    let mut fail = Problem::new(
                        &self.name(),
                        "notdef-produced",
                        format!("Shaper produced a .notdef while {}", string),
                    );
                    if let Some(input_codepoint) = string.char_at(codepoint.cluster as usize) {
                        fail.fixes = vec![Fix {
                            fix_type: "add_codepoint".to_string(),
                            fix_thing: input_codepoint.to_string(),
                        }];
                    }
                    problems.push(fail);
                }
                if checker
                    .codepoint_for(codepoint.glyph_id)
                    .map(simple_mark_check)
                    .unwrap_or(false)
                {
                    if previous.is_some() && previous == dotted_circle && !literally_a_dotted_circle
                    {
                        let fail = Problem {
                            check_name: self.name(),
                            message: format!("Shaper produced a dotted circle when {}", string),
                            code: "dotted-circle-produced".to_string(),
                            terminal: false,
                            context: serde_json::json!({
                                "text": previous,
                                "mark": codepoint.glyph_id,
                            }),
                            fixes: vec![Fix {
                                fix_type: "add_feature".to_string(),
                                fix_thing: format!("to avoid a dotted circle while {}", string),
                            }],
                        };
                        problems.push(fail);
                    } else if position.x_offset == 0 && position.y_offset == 0 {
                        // Suspicious
                        let previous_name = previous.map_or_else(
                            || "the base glyph".to_string(),
                            |gid| {
                                checker
                                    .glyph_names
                                    .get(gid as usize)
                                    .cloned()
                                    .unwrap_or_else(|| format!("Glyph #{}", gid))
                            },
                        );
                        let this_name = checker
                            .glyph_names
                            .get(codepoint.glyph_id as usize)
                            .cloned()
                            .unwrap_or_else(|| format!("Glyph #{}", codepoint.glyph_id));
                        let fail = Problem {
                            check_name: self.name(),
                            terminal: false,
                            message: format!(
                                "Shaper didn't attach {} to {} when {}",
                                this_name, previous_name, string
                            ),
                            code: "orphaned-mark".to_string(),
                            context: serde_json::json!({
                                "text": string,
                                "mark": this_name,
                                "base": previous_name,
                            }),
                            fixes: vec![Fix {
                                fix_type: "add_anchor".to_string(),
                                fix_thing: format!("{}/{}", previous_name, this_name),
                            }],
                        };
                        problems.push(fail);
                    }
                }
                previous = Some(codepoint.glyph_id);
            }
        }
        (problems, tests_run)
    }

    fn describe(&self) -> String {
        format!(
            "Checks that, when {}, no marks are left unattached",
            self.test_strings.iter().map(|x| x.describe()).join(" and ")
        )
    }
}

/// Check if a codepoint is a nonspacing mark
fn simple_mark_check(c: u32) -> bool {
    char::from_u32(c)
        .map(|c| matches!(c.general_category(), GeneralCategory::NonspacingMark))
        .unwrap_or(false)
}

impl NoOrphanedMarks {
    /// Create a new `NoOrphanedMarks` check implementation
    pub fn new(test_strings: Vec<ShapingInput>, has_orthography: bool) -> Self {
        Self {
            test_strings,
            has_orthography,
        }
    }
}
